Migrate start-notebook.sh to bash

yuvipanda opened the pull request:
Based on

> Stop using bash, haha 👍

from https://github.com/jupyter/docker-stacks/issues/1532.

If there's more apetite for this, I'll try to migrate `start.sh` and `start-singleuser.sh` as well - I think they should all be merged together. We can remove the `.sh` suffixes for accuracy, and keep symlinks in so old config still works. Since the shebang is what is used to launch the correct interpreter, the `.sh` doesn't matter.

Will help fix https://github.com/jupyter/docker-stacks/issues/1532, as I believe all those things are going to be easier to do from python than bash

## Describe your changes

## Issue ticket if applicable

<!-- Example - Fix: https://github.com/jupyter/docker-stacks/issues/0 -->

## Checklist (especially for first-time contributors)

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [ ] I will try not to use force-push to make the review process easier for reviewers
- [ ] I have updated the documentation for significant changes

<!-- markdownlint-disable-file MD041 -->

yuvipanda commented:
I'll take a look at the test failures if this kinda migration is acceptable! I'm sure I can sort those out if needed.

mathbunnyru commented:
I like this PR a lot, just a few thoughts about the implementation:
1. Let's add the `.py` extension to the added start-like scripts. It doesn't change much, but some utils (maybe pre-commit hooks) and editors definitely work better (at least not worse).
2. Instead of symlinking to keep compatibility, I propose having simple shell files, which first echo a warning that calling is file is deprecated and then simply pass all the params to the corresponding `.py` file. This way we will be able to first replace a warning with a hard error in a few months and then completely delete the `.sh` files.
3. Let's also document the warning about calling old `.sh` start scripts.
4. I really hope you will have the courage to deal with `start.sh` when we merge this PR 🙂 

Overall, great PR, thank you!

mathbunnyru commented on images/base-notebook/start-notebook:
```diff
@@ -0,0 +1,38 @@
+import sys
+
+# If we are in a JupyterHub, we pass on to `start-singleuser` instead so it does the right thing
+if "JUPYTERHUB_API_TOKEN" not in os.environ:
```
```suggestion
if "JUPYTERHUB_API_TOKEN" in os.environ:
```

mathbunnyru commented on images/base-notebook/start-notebook:
```diff
@@ -0,0 +1,38 @@
+
+# If we are in a JupyterHub, we pass on to `start-singleuser` instead so it does the right thing
+if "JUPYTERHUB_API_TOKEN" not in os.environ:
+    print(
```
You can change this here to `logging.warning`, or in a separate PR (as you wish).

yuvipanda replied:
@mathbunnyru yeah, I think that can be handled separately so this becomes more of a 'straight port'.

mathbunnyru commented:
I updated your branch (to make sure we test latest versions of everything).

yuvipanda commented:
Glad to hear you like it, @mathbunnyru! I've addressed all the points you mentioned and I think this is ready for review now.

I do personally prefer leaving out the `.py` suffix, but am alright leaving it in. I find that most tools are ok looking at the shebang for detection in those cases, but I'm sure that's not 100% of them.

I'm definitely interested in tackling `start.sh` as well! I actually wrote https://github.com/yuvipanda/jupyterhub-roothooks/ a while ago that provides some part of that functionality, so there's gonna be stuff for me to steal from there. I also spent today picking up a plan to replicate as exactly as possible bash's `source` semantics in python and I think I've a path there.

My goal would be to do this in such a way that all the existing tests pass, so no new functionality.

yuvipanda commented:
@mathbunnyru hah, we clashed as I was just about to rebase :)

mathbunnyru commented on docs/using/selecting.md:
````diff
@@ -56,10 +56,17 @@ It contains:
 - Options for a self-signed HTTPS certificate
 
+```{warning}
+`jupyter/base-notebook` also contains `start-notebook.sh` and `start-singleuser.sh` files to maintain backwards
+compatibility. External config that explicitly refers to those files should instead update to refer to
+`start-notebook.py` and `start-singleuser.py` instead. The shim `.sh` files will be removed at some future
+date.
````
```suggestion
`jupyter/base-notebook` also contains `start-notebook.sh` and `start-singleuser.sh` files to maintain backwards compatibility.
External config that explicitly refers to those files should instead
update to refer to `start-notebook.py` and `start-singleuser.py`.
The shim `.sh` files will be removed at some future date.
```

In most of the mardown files here, we try to fit the sentence on one line, and it it's too long, I try to divide on several lines logically (makes it easier to read the source markdown file).

mathbunnyru commented on images/base-notebook/start-notebook.py:
```diff
@@ -0,0 +1,41 @@
+import shlex
+import sys
+
+# If we are in a JupyterHub, we pass on to `start-singleuser.py` instead so it does the right thing
```
I appreciate it that you added comments to this file! I'm sure it will help future contributors to understand this file faster.

yuvipanda replied:
@mathbunnyru you're welcome :)

manics commented:
> I'm definitely interested in tackling start.sh as well! I actually wrote https://github.com/yuvipanda/jupyterhub-roothooks/ a while ago that provides some part of that functionality, so there's gonna be stuff for me to steal from there. I also spent today picking up a plan to replicate as exactly as possible bash's source semantics in python and I think I've a path there.

If you're modifying the behaviour of the start scripts https://github.com/jupyter/docker-stacks/issues/1528 might fit in too.

yuvipanda commented:
Yay, thank you for the quick review and merge, @mathbunnyru. And for the long set of tests that made this easy
